crates/network-scanner-net/src/async_io.rs

```diff
@@ -0,0 +1,152 @@
+        self.poller.delete(socket)?;
+        Ok(())
+    }
+
+    fn start_register_loop(
+        self: Arc<Self>,
+        receiver: Receiver<(Event, std::task::Waker, Arc<Socket>)>,
+    ) -> anyhow::Result<()> {
+        std::thread::Builder::new()
+            .name("[raw-socket]:register-loop | ".to_string())
+            .spawn(move || {
+                tracing::debug!("starting event register loop");
+                loop {
+                    let (event, waker, socket) = match receiver.recv() {
+                        //recv is blocking if channel is empty
+                        Ok(a) => a,
+                        Err(_) => break,
+                    };
+
+                    {
+                        tracing::trace!(?event, ?socket, "registering event");
+                        let mut map = self.map.lock();
+                        if map.contains_key(&event.key) {
+                            continue;
+                        }
+                        map.insert(event.key, waker);
+                    } // drop the lock before registering the event
+
+                    self.poller
+                        .modify(socket, event)
+                        .map_err(|e| tracing::warn!(error = ?e, "Event registration failed"))
+                        .ok(); // cannot handle this error
+                    tracing::trace!("event registered successfully");
+                }
+                tracing::warn!("register loop terminated")
+            })?;
+        Ok(())
+    }
+
+    fn start_loop(self: Arc<Self>) -> anyhow::Result<()> {
+        std::thread::Builder::new()
+            .name("[raw-socket]:io-event-loop | ".to_string())
+            .spawn(move || {
+                let mut events = Events::with_capacity(NonZeroUsize::new(1024).unwrap());
+                tracing::debug!("starting io event loop");
+                loop {
+                    if self.is_terminated.load(std::sync::atomic::Ordering::Acquire) {
+                        break;
+                    }
+
+                    tracing::debug!("polling events");
+                    events.clear();
+                    self.poller.wait(&mut events, None).expect("polling failed");
+                    {
+                        // Important: lock after blocking poll
+                        let mut map = self.map.lock();
+                        for event in events.iter() {
+                            tracing::trace!(?event, "events polled");
+                            let key = event.key;
+                            if let Some(waker) = map.get(&key) {
+                                waker.wake_by_ref();
+                                tracing::trace!(?key, "waker called");
+                                map.remove(&key);
+                            }
+                        }
+                    }
+                }
+                self.is_terminated.store(true, std::sync::atomic::Ordering::SeqCst);
```

Comment:
updated

Reply:
If the thread panics because of the `expect`, then the atomic will not be updated. Did you mean this?

```rust
if let Some(err) = self.poller.wait(&mut events, None) {
    error!(…);
    break;
}
```

Reply:
updated